src/gpgmm/common/CPUPoolAllocator.h

```diff
@@ -0,0 +1,93 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GPGMM_COMMON_CPUPOOLALLOCATOR_H_
+#define GPGMM_COMMON_CPUPOOLALLOCATOR_H_
+
+#include "gpgmm/utils/Assert.h"
+
+#include <queue>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+using AllocationAddress = std::pair<uint64_t, uint64_t>;
+
+namespace gpgmm {
+    template <typename T>
+    class CPUPoolAllocator {
+      public:
+        explicit CPUPoolAllocator(size_t blockSize, size_t objectSize) {
+            mBlockSize = blockSize;
+            mObjectsPerBlock = blockSize / objectSize;
+        }
+
+        T* Allocate(T object) {
+            // If a free allocation is available, reuse that location.
+            if (!mFreeQueue.empty()) {
+                AllocationAddress freeSpot = mFreeQueue.front();
+                mFreeQueue.pop();
+                mAllocations[freeSpot.first][freeSpot.second] = object;
+                return &mAllocations[freeSpot.first][freeSpot.second];
+            }
+
+            // If our next allocation offset starts with a zero it means we haven't allocated the
+            // block yet. Allocate it.
+            if (mNextAllocation.second == 0) {
+                mAllocations.push_back(std::vector<T>());
+                mAllocations[mNextAllocation.first].reserve(mObjectsPerBlock);
+            }
+
+            mAllocations[mNextAllocation.first].push_back(object);
+
+            T* allocationAddress = &mAllocations[mNextAllocation.first][mNextAllocation.second];
+
+            // Map the pointer address to the AllocationAddress so we can keep track of this
+            // location for deallocation.
+            mAllocationAddressMap.insert(
+                {allocationAddress, {mNextAllocation.first, mNextAllocation.second}});
+
+            // Increment next allocation position.
+            mNextAllocation.second++;
+            if (mNextAllocation.second == mObjectsPerBlock) {
+                mNextAllocation.first++;
+                mNextAllocation.second = 0;
+            }
+
+            return allocationAddress;
+        }
+
+        void Deallocate(T* objectPointer) {
+            // Translate T* pointer back to an AllocationAddress
+            typename std::unordered_map<T*, AllocationAddress>::const_iterator search =
+                mAllocationAddressMap.find(objectPointer);
+            ASSERT(search != mAllocationAddressMap.end());
+            AllocationAddress address = search->second;
+            // Insert the AllocationAddress into mFreeQueue
+            mFreeQueue.push(address);
+        }
+
+      private:
+        size_t mBlockSize;
+        uint64_t mObjectsPerBlock;
+
+        AllocationAddress mNextAllocation = {0, 0};
+
+        std::vector<std::vector<T>> mAllocations;
```

Comment:
These STL structures are doing the actual allocation via the default `std::allocator` provided by libc++ (your code is not). This code needs to actually malloc something to guarentee anything will be pooled.

Reply:
I see. To explain further I was trying to work within the awkward constraints I had gotten Udaya to agree to, which is this vector-based pseudo-allocation system. I had brought up DIY-ing an allocator multiple times, however he did not seem to be receptive to the idea. I agree that malloc-ing a chunk of memory is the obvious way to write an allocator, but just as an FYI - this diverges from what Udaya had signed off on. He did not elaborate on why he didn’t agree to a DIY allocator, but I’m guessing he had a preconceived idea that std::pmr could provide the best solution.

Reply:
The underlying elements of a vector are required to be contiguous, so I don't think I understand the issue here.

[1] https://cplusplus.github.io/LWG/issue69

Reply:
...or are you implying that all of this (the nested vectors) will be reallocated upon the containing vector being resized?

Reply:
Managing chunks of memory isn't a feature of using std::vector.  There are no guarentees std::vector will malloc (ex. resize) that (contiguous) data from the same chunk.

Reply:
This is changed to something like `std::vector<std::vector<T>*> mAllocations;`. The memory of any underlying nested vectors should remain fixed for the life of the object because we do `.reserve(mNumObjectsPerBlock)` at creation time and we never change the size of the vector. Reallocation of the outside vector shouldn't matter because it only holds pointers to the fixed blocks.

Reply:
But the inside vector itself still must malloc somewhere on the heap, no? Either you create a massive std::vector via mNumObjectsPerBlock (not really practical) or create a pool-allocator to reuse their memory.